File: mutual-exclusion/me.java
// package mutual_exclusion;
import java.util.*;
public class me 
{
    public static void main(String[] args) 
    {
        int ch,n,co,i,j,cs,critical = 0;
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter a no. processes :");
        n = sc.nextInt();
        critical = n;
        ArrayList process = new ArrayList();
        ArrayList wait = new ArrayList();

        System.out.println("Enter processess :");
        for( i=0;i<n;i++)
        {
            if(i == 0)
            {
                System.out.println("CRITICAL SECTION is EMPTY");
                j = sc.nextInt();
                process.add(j);
                System.out.println("process goes into CS..");
            }
            else
            {
                j = sc.nextInt();
                wait.add(j);
                System.out.println("Process goes into waiting queue");
            }
        }
//        System.out.println("Enter process co-ordinator number :");
//        co = sc.nextInt();
        do
        {
            System.out.println("");
            System.out.println("**** MENU ****");
            System.out.println("1.ADD");
            System.out.println("2.REMOVE");
            System.out.println("3.VIEW SYSTEM STATUS");
            System.out.println("99.Exit");
            System.out.println("Enter your choice :");
            ch = sc.nextInt();
            
            int p;
                        
            switch(ch)
            {
                case 1:
                    {
                        System.out.println("ADD process");
                        p = sc.nextInt();
                        if(critical == n)
                        {
                            System.out.println("wait ..");
                            wait.add(p);
                            System.out.println("Added to waiting queue");
                        }
                        else
                        {
                            process.add(p);
                            System.out.println("Process added...");
                            critical++;
                        }
                        break;
                    }
                case 2:
                    {
                        if(critical == 0)
                        {
                            System.out.println("critical section is Empty");
                            //wait.add(p);
                        }
                        else
                        {
                            System.out.println("REMOVE");
                            process.remove(0);
                            System.out.println("resource released ...");
                            critical--;
                        }
                            
                        break;

                    }
                case 3:
                    {
                        System.out.println("System status");
                        Iterator it = process.iterator();
                        while(it.hasNext())
                        {
                            System.out.print(it.next()+"\t");
                        }
                        break;
                    }
                
                case 99:
                    {
                        System.out.println("Thank you.....");
                        break;
                    }
                default:
                    {
                        System.out.println("Invalid Input");
                        break;
                    }
                
            }
            
        }while(ch != 99 );
    }
}

File: mutual-exclusion/me.cpp
#include <iostream>
using namespace std;
int i, n, queue[10], crit = 0;
void add();
void release();
void status();
int main()
{
    int process[10];
    printf("Enter number of process: (max 10)");
    cout << endl;
    scanf("%d", &n);
    int a = 1;
    if (n > 10)
    {
        printf("invalid number");
    }
    else
    {
        for (i = 0; i < n; i++)
        {
            process[i] = a;
            a++;
        }
    }

    for (i = 0; i < 10; i++)
    {
        queue[i] = 0;
    }

    int x = 0;
    while (x == 0)
    {
        int choice;
        printf("MENU\n");
        printf("1: Request\n");
        printf("2: Release\n");
        printf("3: Exit\n");
        printf("4: Status");
        cout << endl;
        scanf("%d", &choice);
        switch (choice)
        {
        case 1:
            add();
            break;
        case 2:
            release();
            break;
        case 3:
            x = 1;
            break;
        case 4:
            status();
            break;
        default:
            printf("Invalid choice");
        }
    }
}
void add()
{
    int p;
    printf("which process you want to add");
    cout << endl;
    scanf("%d", &p);

    if (crit == 0)
    {
        crit = p;
    }
    else
    {
        for (i = 0; i < 9; i++)
        {
            if (queue[i] == 0)
            {
                queue[i] = p;
                break;
            }
        }
    }
}
void release()
{
    if (crit == 0)
    {
        printf("Critical section is already empty\n");
    }
    else
    {
        crit = 0;
        if (queue[0] != 0)
            crit = queue[0];
        for (i = 0; i < 3; i++)
        {
            queue[i] = queue[i + 1];
        }
        printf("%d is accessing crit section", crit);
        cout << endl;
    }
}
void status()
{
    if (crit == 0)
    {
        printf("Critical Section is currently FREE\n");
    }
    else
    {
        printf("Process %d is accessing critical section\n", crit);
    }
    for (i = 0; i < n; i++)
    {
        printf(" ");
        printf("%d", queue[i]);
    }
}